core/net/rpl/rpl-icmp6.c

```diff
@@ -885,13 +887,148 @@ dao_input(void)
+
+  buffer = UIP_ICMP_PAYLOAD;
+  buffer_length = uip_len - uip_l3_icmp_hdr_len;
+
+  pos = 0;
+  instance_id = buffer[pos++];
+  instance = rpl_get_instance(instance_id);
+  lifetime = instance->default_lifetime;
+
+  flags = buffer[pos++];
+  /* reserved */
+  pos++;
+  sequence = buffer[pos++];
+
+  dag = instance->current_dag;
+  /* Is the DAG ID present? */
+  if(flags & RPL_DAO_D_FLAG) {
+    if(memcmp(&dag->dag_id, &buffer[pos], sizeof(dag->dag_id))) {
+      PRINTF("RPL: Ignoring a DAO for a DAG different from ours\n");
+      return;
+    }
+    pos += 16;
+  }
+
+  /* Check if there are any RPL options present. */
+  for(i = pos; i < buffer_length; i += len) {
+    subopt_type = buffer[i];
+    if(subopt_type == RPL_OPTION_PAD1) {
+      len = 1;
+    } else {
+      /* The option consists of a two-byte header and a payload. */
+      len = 2 + buffer[i + 1];
+    }
+
+    switch(subopt_type) {
+    case RPL_OPTION_TARGET:
+      /* Handle the target option. */
+      prefixlen = buffer[i + 3];
+      memset(&prefix, 0, sizeof(prefix));
+      memcpy(&prefix, buffer + i + 4, (prefixlen + 7) / CHAR_BIT);
+      break;
+    case RPL_OPTION_TRANSIT:
+      /* The path sequence and control are ignored. */
+      /*      pathcontrol = buffer[i + 3];
+              pathsequence = buffer[i + 4];*/
+      lifetime = buffer[i + 5];
+      if(len >= 20) {
+        memcpy(&dao_parent_addr, buffer + i + 6, 16);
+      }
+      break;
+    }
+  }
+
+  PRINTF("RPL: DAO lifetime: %u, prefix length: %u prefix: ",
+          (unsigned)lifetime, (unsigned)prefixlen);
+  PRINT6ADDR(&prefix);
+  PRINTF(", parent: ");
+  PRINT6ADDR(&dao_parent_addr);
+  PRINTF(" \n");
+
+  if(lifetime == RPL_ZERO_LIFETIME) {
+    PRINTF("RPL: No-Path DAO received\n");
+    rpl_ns_expire_parent(dag, &prefix, &dao_parent_addr);
+  } else {
+    if(rpl_ns_update_node(dag, &prefix, &dao_parent_addr, RPL_LIFETIME(instance, lifetime)) == NULL) {
+      PRINTF("RPL: failed to add link\n");
+      return;
+    }
+  }
+
+  if(flags & RPL_DAO_K_FLAG) {
+    PRINTF("RPL: Sending DAO ACK\n");
+    uip_clear_buf();
+    dao_ack_output(instance, &dao_sender_addr, sequence,
+        RPL_DAO_ACK_UNCONDITIONAL_ACCEPT);
+  }
+#endif /* RPL_WITH_NON_STORING */
+}
+/*---------------------------------------------------------------------------*/
+static void
+dao_input(void)
+{
+  rpl_instance_t *instance;
+  uint8_t instance_id;
+
+  /* Destination Advertisement Object */
+  PRINTF("RPL: Received a DAO from ");
+  PRINT6ADDR(&UIP_IP_BUF->srcipaddr);
+  PRINTF("\n");
+
+  instance_id = UIP_ICMP_PAYLOAD[0];
+  instance = rpl_get_instance(instance_id);
+  if(instance == NULL) {
+    PRINTF("RPL: Ignoring a DAO for an unknown RPL instance(%u)\n",
+           instance_id);
+    goto discard;
+  }
+
+  if(instance->mop != RPL_MOP_NON_STORING) {
+    if(RPL_IS_STORING(instance)) {
+      dao_input_storing();
+    }
+  } else {
+    if(RPL_IS_NON_STORING(instance)) {
```

Comment:
Do you need to check instance->mop both in the top-level <code>if</code> statement and in the nested if statement? Perhaps it would also be good to add a debug statement if we receive an unsupported MOP that is discarded.